### PR TITLE
Rename Mesh::GetBdrElementAdjacentElement2 to Mesh::GetBdrElementAdjacentElementWithInverseOrientation

### LINKED ISSUE #3879: Possible typo in GetBdrElementAdjacentElement2(...)
Hi, I've noticed that the `GetBdrElementAdjacentElement2()` function in [mesh.cpp](https://github.com/mfem/mfem/blob/master/mesh/mesh.cpp) returns `Elem1No` as opposed to `Elem2No` of the boundary element in question. 

I imagine this is a typo since at the moment the two functions `GetBdrElementAdjacentElement()` and `GetBdrElementAdjacentElement2()` return exactly the same result (with the exception of the Triangle and Square geometries, where the orientation is reversed, if I understand things correctly.

Now this of course wouldn't be relevant for external boundaries, where the `Elem2No` returned will just be -1, but in cases where one needs to loop over internal boundary attributes and recover the elements on either side, it does become important.

Let me know if indeed this is a typo or if perhaps I've missed something. In case it is a typo, it should be a very quick fix, just changing `Elem1No` to `Elem2No` in the `GetBdrElementAdjacentElement2()` method.

### LINKED ISSUE #3985: Bug in Mesh::GetBdrElementAdjacentElement2
There are two distinct mesh functions that appear to be completely identical:
Mesh::GetBdrElementAdjacentElement and Mesh::GetBdrElementAdjacentElement2 seem to do the exact same thing.  I think the last two lines of Mesh::GetBdrElementAdjacentElement2 should be changed from:
`el   = fi.Elem1No;
info = fi.Elem1Inf + ori;`
to
`el   = fi.Elem2No;
info = fi.Elem2Inf + ori;`

### PR DESCRIPTION
**Pull request by pazner**
Resolves #3879 (and resolves #3985).

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4034](https://github.com/mfem/mfem/pull/4034) | @pazner | @tzanio | @tzanio + @v-dobrev | 12/27/23 | 2/13/24 | 2/19/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by tzanio — APPROVED** [review]
Not a huge fan of such a long function names, but couldn't come up with a better suggestion.

**Comment by pazner** [comment]
> Not a huge fan of such a long function names, but couldn't come up with a better suggestion.

I'm definitely open to alternatives.

One option: do a bit of a refactor, so that only we only have `Mesh::GetBdrElementAdjacentElement`. We can add a helper function that, given an orientation, returns the inverted orientation (this is just indexing into `Geometry::Constants<geom_t>::InvOrient`), and then an additional helper function that encodes and decodes the `info` integer (this is probably overdue anyway).

That way, we can just get rid of the inverted accessor.

**Comment by pazner** [comment]
Alternative solution [here](https://github.com/mfem/mfem/compare/getbdrelementadjacentelement2-refactor)

**Comment by tzanio** [comment]
> Alternative solution [here](https://github.com/mfem/mfem/compare/getbdrelementadjacentelement2-refactor)

I personally like this ☝️ much more!

**Comment by pazner** [comment]
> > Alternative solution [here](https://github.com/mfem/mfem/compare/getbdrelementadjacentelement2-refactor)
> 
> I personally like this ☝️ much more!

@v-dobrev, what do you think? Should I merge the alternative suggestion here?

**Review by Heinrich-BR — COMMENTED** [review]
Simply renaming the function to `GetBdrElementAdjacentElementWithInverseOrientation` certainly does the trick, but as @tzanio has pointed out, the name is now too long. Also, having a whole function just to return the same face info but with inverse orientation feels like unnecessary bloat. I also prefer the [alternative solution](https://github.com/mfem/mfem/compare/getbdrelementadjacentelement2-refactor) given by @pazner which makes the orientation reversal explicit. Would be happy to merge that into this PR.

**Comment by tzanio** [comment]
ping: @v-dobrev

**Review by v-dobrev — APPROVED** [review]

**Comment by v-dobrev** [comment]
Oops, I now see that there was a proposed alternative solution and I already approved.

Taking a look at the proposed alternative ...

**Comment by v-dobrev** [comment]
The alternative looks okay to me too. I'm not sure I like it better but that's fine.

One request: there are some long lines in the alternative that need to be reformatted.

**Comment by pazner** [comment]
Thanks for taking a look @v-dobrev. I merged the alternative solution (and master) into this branch.

File: mesh/mesh.hpp [hunk 1]
```diff
@@ -1944,6 +1954,17 @@ class Mesh
       operator Mesh::FaceInfo() const;
    };
 
+   /// Given a "face info int", return the face orientation. @sa FaceInfo.
+   static int DecodeFaceInfoOrientaiton(int info) { return info%64; }
```
Comment (vladotomov) [inline mesh/mesh.hpp:1958]: typo `orientaiton`

Comment (pazner) [reply mesh/mesh.hpp]: Oops, thanks for pointing it out. Fixed in 76006e9b1.

File: fem/geom.cpp [hunk 2]
```diff
@@ -252,6 +252,33 @@ Geometry::Geometry()
    }
 }
 
+template <Geometry::Type GEOM>
+int GetInverseOrientation_(int orientation)
+{
+   using geom_t = Geometry::Constants<GEOM>;
+   MFEM_ASSERT(orientation < geom_t::NumOrient, "Invalid orientation");
```
Comment (v-dobrev) [inline fem/geom.cpp:259]:
```suggestion
   MFEM_ASSERT(0 <= orientation && orientation < geom_t::NumOrient,
               "Invalid orientation");
```

File: mesh/submesh/submesh_utils.cpp [hunk 3]
```diff
@@ -117,18 +117,21 @@ void BuildVdofToVdofMap(const FiniteElementSpace& subfes,
 
             auto pm = parentfes.GetMesh();
 
+            const Geometry::Type face_geom =
+               pm->GetBdrElementBaseGeometry(parent_element_ids[i]);
```
Comment (v-dobrev) [inline mesh/submesh/submesh_utils.cpp:121]: Similar to above, this call can also be replaced by a call to `GetBdrElementGeometry`.

Comment (pazner) [reply mesh/submesh/submesh_utils.cpp]: I went ahead and replaced all calls to `Mesh::GetBdrElementBaseGeometry` with `Mesh::GetBdrElementGeometry` in a67d9f915.

File: mesh/mesh.hpp [hunk 4]
```diff
@@ -1467,15 +1467,25 @@ class Mesh
...
        the boundary element relative to the respective face element. In other
        words this is the orientation of the face element relative to the
        boundary element.
 
+       @warning This only differs from GetBdrElementAdjacentElement by returning
+       the face info with inverted orientation. It does @b not return
+       information corresponding to a second adjacent face. This function is
+       deprecated, use Geometry::GetInverseOrientation, Mesh::EncodeFaceInfo,
+       Mesh::DecodeFaceInfoOrientation, and Mesh::DecodeFaceInfoLocalIndex
+       instead.
```
[the remark this answers is not in the document]
Comment (pazner) [reply mesh/mesh.hpp:1485]: My personal opinion is that it's used in so few places in the library (just two places if I am counting correctly), and likely will not be used often in user application, so it's not a big deal that users have to jump through a few hoops to do the same thing.

[the remark this answers is not in the document]
Comment (tzanio) [reply mesh/mesh.hpp:1485]: I personally agree with @pazner

**Review by v-dobrev — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**